### DISCUSSION TITLE
How do I create callbacks like OnCollisionEnter, OnCollisionExit, OnCollisionStay?

### DISCUSSION THREAD

**Discussion by OhJeongrok** [notgiven688/jitterphysics2#69] [How do I ... ?]
I would like to create a function that creates things like trigger volumes in the world and handles these events. How can I do this?

Should I implement it by using Filter(BroadPhaseFilter, NarrowPhaseFilter) and managing the conflicted list?

**Comment by notgiven688** [comment]
Jitter does not provide these methods for you, but you are free to build your own system.

```
    private Shape testSphere = new SphereShape(1);
    private JVector testSpherePosition = new JVector(0, 4, 0);
    
    private List<Shape> potentialCollisions = new List<Shape>();

    void UpdateTriggerVolumeTest()
    {
        testSphere.CalculateBoundingBox(JMatrix.Identity, testSpherePosition, out JBBox boxToTest);
        world.DynamicTree.Query(potentialCollisions, boxToTest);

        foreach (var shape in potentialCollisions)
        {
            RigidBody? rb = shape.RigidBody;
            
            if (rb == null) continue;

            bool hit = NarrowPhase.MPREPA(shape, testSphere,
                rb.Orientation, JMatrix.Identity,
                rb.Position, testSpherePosition,
                out JVector pointA, out JVector pointB, out JVector normal, out float penetration);

            // hit, pointA, pointB, normal, penetration are available at this point
        }
        
        potentialCollisions.Clear();
    }
```

This code should be quite efficient. It calculates the bounding box for the testSphere and queries the collision system for potential overlaps of other shapes. With a call to NarrowPhase.MPREPA you can get detailed collision information.

**Comment by notgiven688** [comment]
If you want to keep track of collisions you can just overwrite the NarrowPhase-Filter  - as you have suggested.
Also keep in mind that Jitter stores the contacts of each body in body.Contacts.
